source/taskmanager/src/main/java/nl/aerius/taskmanager/mq/RabbitMQWorkerSizeProvider.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright the State of the Netherlands
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+package nl.aerius.taskmanager.mq;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import nl.aerius.taskmanager.adaptor.WorkerProducer.WorkerMetrics;
+import nl.aerius.taskmanager.adaptor.WorkerSizeObserver;
+import nl.aerius.taskmanager.adaptor.WorkerSizeProviderProxy;
+import nl.aerius.taskmanager.client.BrokerConnectionFactory;
+
+/**
+ * Provider to use different means to get information about the size and utilisation of the workers.
+ * It both queries the RabbitMQ admin interface to get a total insight as it listens to changes in the number of consumers registering
+ * with RabbitMQ.
```

Comment:
So it can work on both the fanout mechanism, which supplies deltas, and the management API mechanism, which supplies totals?

Won't that cause possible race conditions? For instance, a worker gets started, and registers to RabbitMQ, and triggers a fanout message. Now when the API management update happens to be triggered right away, it'll update to the new total right away. If then the fanout message gets received, it'll update the new total with the delta (+1), and you end up with an extra worker in your local state (which gets reset again with the next API management update I guess). 

Guess if the API management call usually lags behind a bit, it's not that big of a deal, just wondering (and I could be missing something).

Reply:
Yes it's possible. We'll have to see if this is causing to much issues. Otherwise we need to see how this can be handled. In the end it should not be much of a problem because the API management update will restore the correct state, just as you mention.

Reply:
I'd prefer these to be mutually exclusive, unless we have a reason not to trust the events plugin. Whether this needs to be configured or maybe a fallback way (try events, if that isn't available use the management API), both are fine I guess. Leaning towards the latter if it can be done cleanly.

My reasons for this are that it would get rid of the required dependency for de management API and make it optional. The less dependencies the better.
Another reason is that most cloud providers have their own RabbitMQ proxy or support the `AQMP` protocol with their hosted message services, so the management API will probably not be present. Not sure if they would support the events plugin either, but yeah.

Reply:
Think the problem with the events mechanism is that it won't be possible to generate an accurate number of current workers based on that alone? If a worker is started at a later time (for instance upscaling number of workers), it'll have missed all the events. Same with a restart of the taskmanager, all events are gone already, so there's no way to tell current number of workers based on events alone.

Could perhaps isolate the dependency in the taskmanager alone and not in the clients if you let taskmanager generate messages about current nr of workers to a fanout exchange that the clients in turn can listen to?

Reply:
Is the only client that needs to know the number of workers the chunker so it can scale and use any unused resources, or are there more use-cases I am forgetting?

Reply:
moved comment:
> I'd prefer these to be mutually exclusive, unless we have a reason not to trust the events plugin.

The events are only triggered when the worker starts. When the taskmanager start later or is restarted it doesn't know the workers so it needs the call to the admin api.

> My reasons for this are that it would get rid of the required dependency for de management API and make it optional. The less dependencies the better.

It's a choice between inventing our own wheel or use what is available. If we don't use these plugings we would have to write something ourselves to get this kind of information. That has not my preference. There is a large dependency on RabbitMQ. I don't know if it would be possible to use another AMPQ (although it's wrapped in interfaces, so another implementation is possible if needed).
With this change the workers won't be dependent on the admin api anymore.

> Is the only client that needs to know the number of workers the chunker so it can scale and use any unused resources, or are there more use-cases I am forgetting?

Both the taskmanager and the clients need to know the number of workers. At this moment there are no other use-cases. But it might be we can use the information published by the taskmanager about the number of workers available could be used to auto-scale.